### PR TITLE
Fix HostPortPair comparison

### PR DESCRIPTION
**Pull request by atuchin-m**
Resolves https://github.com/brave/brave-browser/issues/36535

The PR:
* fixes the crash ([the details](https://github.com/brave/brave-browser/issues/36535#issuecomment-2163761449) in the issue)
* simplify the comparison code to prevent future issues.
* fixes another issue with parsing `HostPortPair`.
* adds unit tests;

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/net/base/host_port_pair.cc [hunk 1]
```diff
@@ -33,7 +33,7 @@ HostPortPair FromStringWithAuthentication(std::string_view str) {
       HostPortPair::FromString(std::string(auth_host[1]));
 
   std::vector<std::string_view> user_pass = base::SplitStringPiece(
-      str, ":", base::TRIM_WHITESPACE, base::SPLIT_WANT_ALL);
+      auth_host[0], ":", base::TRIM_WHITESPACE, base::SPLIT_WANT_ALL);
```
Comment (atuchin-m) [inline chromium_src/net/base/host_port_pair.cc:36]: It was another bug revealed by `BraveHostPortPairTest.Parsing` test

File: chromium_src/net/base/host_port_pair.cc [hunk 2]
```diff
@@ -138,21 +130,22 @@ void HostPortPair::set_password(const std::string& in_password) {
   password_ = in_password;
 }
 
-bool HostPortPair::operator<(const HostPortPair& other) const {
-  if (HostPortPair_ChromiumImpl::operator<(other)) {
-    return true;
-  }
-  return std::tie(username_, password_) <
```
Comment (atuchin-m) [inline chromium_src/net/base/host_port_pair.cc:145]: Here is the main problem: if `*this > other ` we should return false, but we fallback to `(username_, password_)` comparison resulting in an inconsistent implementation.

**Review by boocmp — APPROVED** [review]

**Review by iefremov — APPROVED** [review]

**Comment by iefremov** [comment]
pls uplift the fix to beta/stable

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win 11 x64` using the following build(s):

```
Brave | 1.69.19 Chromium: 126.0.6478.56 (Official Build) nightly (64-bit)
-- | --
Revision | f949b8805f9aed30d2cd2fce2350261918ffa0c9
OS | Windows 11 Version 23H2 (Build 22631.3737)
```

Using the STR/Cases outlined via https://github.com/brave/brave-core/pull/22053#issuecomment-1957706020 as per https://github.com/brave/brave-core/pull/24167#issue-2349511284, ensured that `Tor` was generally working as per the following:

* ensured that opening a `Tor` window worked without any issues/failures
* ensured that `Tor` connected without any issues/failures
* visited https://check.torproject.org and ensured that a new IP is being assigned every time when `Ctrl+Shift+R/Cmd+Shift+R`
  * ensured that a new IP is being assigned every time when via `New Tor connection for this site`
 
**`Example`** | **`Example`** | **`Example`**
----------------|----------------|-----------------
image | image | image

**Pull request merged**